Boost.Regex/ICU will no longer compile with clang 19

This simple test program triggers a warning in clang 18 that indicates that the code will no longer compile with clang 19:
```c++
#include <iostream>
#include <boost/regex/icu.hpp>

int main()
{
    const boost::u32regex r1(boost::make_u32regex(".*"));
    const boost::u32regex r2(boost::make_u32regex(".*"));

    std::cout << "identical: " << (r1==r2?"yes":"no") << std::endl;
    return 0;
}
```

> In file included from main.cpp:1:
In file included from include/c++/v1/iostream:43:
In file included from include/c++/v1/ios:221:
In file included from include/c++/v1/__fwd/ios.h:13:
include/c++/v1/__fwd/string.h:45:41: warning: 'char_traits<int>' is deprecated: char_traits<T> for T not equal to char, wchar_t, char8_t, char16_t or char32_t is non-standard and is provided for a temporary period. It will be removed in LLVM 19, so please migrate off of it. [-Wdeprecated-declarations]
   45 | template <class _CharT, class _Traits = char_traits<_CharT>, class _Allocator = allocator<_CharT> >
      |                                         ^
Boost/1.83.0/include/boost/regex/v5/basic_regex.hpp:237:24: note: in instantiation of default argument for 'basic_string<int>' required here
  237 |          result = std::basic_string<charT>(this->m_expression, this->m_expression_len);
      |                        ^~~~~~~~~~~~~~~~~~~
Boost/1.83.0/include/boost/regex/v5/basic_regex.hpp:466:39: note: in instantiation of member function 'boost::re_detail_500::basic_regex_implementation<int, boost::icu_regex_traits>::str' requested here
  466 |       return m_pimpl.get() ? m_pimpl->str() : std::basic_string<charT>();
      |                                       ^
Boost/1.83.0/include/boost/regex/v5/basic_regex.hpp:536:14: note: in instantiation of member function 'boost::basic_regex<int, boost::icu_regex_traits>::str' requested here
  536 |       return str().compare(that.str());
      |              ^
Boost/1.83.0/include/boost/regex/v5/basic_regex.hpp:540:14: note: in instantiation of member function 'boost::basic_regex<int, boost::icu_regex_traits>::compare' requested here
  540 |       return compare(e) == 0;
      |              ^
main.cpp:9:38: note: in instantiation of member function 'boost::basic_regex<int, boost::icu_regex_traits>::operator==' requested here

The problem seems to be the character type UChar32 (which is actually int32_t) when Boost.Regex 1.83.0 is used in combination with the ICU library.

--- a/include/boost/regex/concepts.hpp
+++ b/include/boost/regex/concepts.hpp
@@ -76,13 +76,10 @@ inline long hash_value(char_architype val)
 //
 } // namespace boost
 namespace std{
-   template<> struct char_traits<boost::char_architype>
-   {
-      // The intent is that this template is not instantiated,
-      // but this typedef gives us a chance of compilation in
-      // case it is:
-      typedef boost::char_architype char_type;
-   };
+   //
+   // We should never use this, if we do it should be an error:
+   //
+   template<> struct char_traits<boost::char_architype>;
 }
 //
 // Allocator architype:
@@ -411,6 +408,10 @@ struct BaseRegexConcept
       ignore_unused_variable_warning(e4);
       Regex e5(in1, in2, m_flags);
       ignore_unused_variable_warning(e5);
+
+      // equals:
+      e1 == e2;
+      e1 != e2;
 
       // assign etc:
       Regex e;

--- a/include/boost/regex/v5/basic_regex.hpp
+++ b/include/boost/regex/v5/basic_regex.hpp
@@ -533,7 +533,21 @@ public:
          return status() - that.status();
       if(flags() != that.flags())
          return flags() - that.flags();
-      return str().compare(that.str());
+      
+      const char_type* i = m_pimpl->begin();
+      const char_type* j = that.m_pimpl->begin();
+      while ((i != m_pimpl->end()) && (j != that.m_pimpl->end()))
+      {
+         if (*i != *j)
+            return *i < *j ? -1 : 1;
+         ++i;
+         ++j;
+      }
+      if (i != m_pimpl->end())
+         return *i > static_cast<char_type>(0) ? 1 : -1;
+      if (j != that.m_pimpl->end())
+         return *j > static_cast<char_type>(0) ? -1 : 1;
+      return 0;
    }
    bool  operator==(const basic_regex& e)const
    { 

